### PR TITLE
DataTable Module

### PR DESCRIPTION
**Pull request by ajkauffmann**
This is a new module that uses the DataTable.Compute method to calculate an expression.

https://docs.microsoft.com/en-us/dotnet/api/system.data.datatable.compute?view=net-6.0

### CODE REVIEW

**Comment by pri-kise** [comment]
@ajkauffmann compared to the other PR, I guess you should add your permissionset to the System Objects Permission Set (https://github.com/microsoft/ALAppExtensions/blob/main/Modules/System/Permissions/SystemApplicationObjects.PermissionSet.al).

**Comment by ajkauffmann** [comment]
@pri-kise good catch, overlooked that one. PR has been updated.

**Comment by pri-kise** [comment]
> @pri-kise good catch, overlooked that one. PR has been updated.
I was keeping track of the Sharepoint PR and did recognize that this extension was missing there, too. :)

I was surprised that you used variants passed by reference. I thought this wasn't possible. (Check PR https://github.com/microsoft/ALAppExtensions/pull/12281)
Is there any documentation, that this behaviour has changed?

And do we still need to explicit cast variables to variants, when using the TryFunction?
https://experience.dynamics.com/ideas/idea/?ideaid=08a6c1b1-2583-eb11-8ced-0003ff45ceba

**Comment by NKarolak** [comment]
`But passing by reference requires the original variable to be of type variant.`

That piece of information was missing for me all these years 😂🙈

**Comment by ajkauffmann** [comment]
@NKarolak Same reason why you can't pass a Label into a by reference text parameter.

File: Modules/System/DataTable/src/DataTableImpl.Codeunit.al [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        DotNetType: DotNet Type;
+        DotNetConvert: DotNet Convert;
+        Integer: Integer;
+        BigInteger: BigInteger;
+        Decimal: Decimal;
+        Boolean: Boolean;
+    begin
+        DotNetType := DotNetResult.GetType();
+        case true of
+            DotNetType.Equals(GetDotNetType(DotNetByte)),
```
[the remark this answers is not in the document]
Comment (ajkauffmann) [reply Modules/System/DataTable/src/DataTableImpl.Codeunit.al:57]: Originally the code could theoretically return a byte. But it seems that compute will never return a byte anyway, so I decided to leave it in there combined with an integer just for safety reasons. Updated the readme.md but forgot about procedure documentation.

Besides, a byte in AL is a numeric value, but we usually don't use it for the same purpose as with integers.

**Comment by nicolassaleron** [comment]
This remark is not particulary about this PR but much more about the module naming convention in the System App.

Does the module name have to stick with the implementation in the .Net framework? If yes, I do not really get the idea around the public and implementation codeunit pattern :D
To use this PR as an example, for instance, computing an expression can be done in different ways : you can also achieve this with XSL transformation today in BC. If the implementation changes in the future (or if there is multiple implementations), will the name DataTable still relevant?

I am also wondering if it would be easy for an AL developer (without .Net background) to get why an expression evaluator API is named DataTable.
However, it is absolutely relevant if there is an intent to expose more features from .Net DataTable object.

Except for the name :), I hope it will be available quickly: we are missing such feature!

**Comment by ajkauffmann** [comment]
@nicolassaleron Very good point, I have been thinking about this as well. For now, I've chosen to follow the other modules that are named after the .Net object. The implementation codeunit is used because there is some logic behind it.

Another option would be to add this to the Math module. That's a module I would be looking at first in case I was searching for calculating an expression. However, that module is almost a 1-to-1 implementation of the .Net Math object. So I wasn't sure if that would be correct or not.

But I agree, you won't look for the DataTable module because of calculating expressions unless you know about the Compute method. But then again, the Compute method is theoretically not meant for such expressions. Its purpose is to create columns and rows with values and then compute an expression using values from the table. Some would even say that using Compute in the way I did is actually abusing it.

I'm not sure about extending it with other DataTable features. I have the feeling that it quickly becomes too complex to handle all kind of different situations. But maybe I'm overthinking it...

**Comment by PhDuck** [comment]
Adding an expression parser and evaluator to the language is a significant change, (ab)using the fact that .NET’s DataTable exposes one is likely a poor choice. AL a statically typed language, where computations are done on top of typed values and their operations are well defined.

Furthermore, the expression syntax supported for DataTable is specialized for DataTable, so the following expression: “2 + 2 + Mads” expects a column in the DataTable named “Mads”. 

This type of dynamic input handling is fundamentally at odds with the statically typed nature of AL. Errors will only happen at runtime and they will be often be very hard to debug. 

Expression parsing and evaluation can be computationally expensive, if a sufficiently complex expression is passed this could serve as DOS target. These calculations are not cancelable leaving us with the possibility of threads that does not interact nicely with user-level scheduling.

In general, the type DataTable is considered legacy in the .NET environment, so I wouldn’t want to rely on its future existence:
> The DataSet and DataTable types are legacy .NET components that allow representing data sets as managed objects.(1)[https://learn.microsoft.com/en-us/dotnet/framework/data/adonet/dataset-datatable-dataview/security-guidance]

**Comment by ajkauffmann** [comment]
@PhDuck Thanks for your reply and thoughts. 
You asked about the key scenario. I've written this for a partner with a solution in project-based production. The solution is configurable and a key part is a feature to declare formulas for the calculation of the most optimized bill of materials. Previously in C/AL this was done by evaluating the formulas with the DataTable object.
Theoretically, in the cloud this could be achieved by moving the code to Azure Functions. However, the vast amount of calls to this feature doesn't really fit with Azure Functions. It's slowing down the performance and increasing costs. The best option would be to have it built into the AL platform.

Now that you expressed your concerns about using DataTable for this (or abusing it, which is in fact true), do you have any alternative? I don't know about any other feature in .Net that does the same. There are plenty of 3rd party libraries, is it an option to select one? I guess that's opening another can of worms, right? I have also looked at writing it in pure AL code, but working with .Net objects in AL is not exactly the same as writing in C#.

**Comment by nicolassaleron** [comment]
@ajkauffmann We have also developed such feature for BC online customers without the DataTable. Actually you can already do this with XSLT transformations. Obviously, the formulas are not exactly the same: you can have a look [here](http://docs.capvision-cloud.fr/fr-FR/corelibrary/parsers.html?tabs=rrp-sample-1%2Ctpp-sample-1%2Ctpp-sample-2%2Cev-sample-1#expression-evaluator) in our (short) documentation (in French).

The key for that is the procedure XmlDomManagement.TryTransformXMLToOutStream: you pass an XML document with your variables, you create an XSL document with your formula and merge them to get a new XML document with the result.

I can share the code in a private message if you want to have a look and give me your opinion if it is worth to create a standard module with it.

**Comment by ajkauffmann** [comment]
@nicolassaleron thanks for your suggestion. In my opinion, XSLT is about manipulating text, not numbers. It works, but also depends on the XSLT processor and how fast and accurate it is. Besides, we are dealing with existing installations with many formulas that would need to be translated into a different formula structure. I think that's not feasible, to be honest.

**Comment by ajkauffmann** [comment]
Ok... I had a brainwave... And I wrote an expression evaluator in pure AL (only used the Stack .Net object for storing intermediate values).

Results:

image

Please come up with a complex formula to test... ☺️

**Comment by JesperSchulz** [comment]
I like brainwaves! Now I'm curious what you did 😎
You should design the module to fulfill _your_ needs. If it does, your job is done! Others can then expand on it, if needed.

**Comment by ajkauffmann** [comment]
@JesperSchulz Yes, let's abandon this PR. I have found another solution that evaluates formulas in AL rather than using (or abusing as some would say) the DataTable object.
I'd be glad to create another PR for it, just wondering if it should be a module in its own or part of the Math module. What do you think?

**Comment by JesperSchulz** [comment]
Almost sounds like a good fit for the math module, doesn't it? Even though Math for now merely is a wrapper for the .NET Math class - but it doesn't have to stay like that! Anyway, closing this PR.

**Pull request closed without merging**